cplusplus/gen-operators.py

```diff
@@ -1,212 +1,257 @@
+    result += '        (options ? options : VImage::option())'
+    if member_x is not None:
+        result += '->\n'
+        result += '            set( "{0}", *this )'.format(member_x)
+
+    all_required = required_input
+
+    if has_output:
+        # first element needs to be passed by reference
+        arg = cppize(required_output[0])
+        result += '->\n'
+        result += '            set( "{0}", &{1} )'.format(required_output[0], arg)
+
+        # append the remaining list
+        all_required += required_output[1:]
+
+    for name in all_required:
+        arg = cppize(name)
+        result += '->\n'
+        result += '            set( "{0}", {1} )'.format(name, arg)
+
+    result += ' );\n'
+
+    if has_output:
+        result += '\n'
+        result += '    return( {0} );\n'.format(required_output[0])
+
+    result += '}\n'
+
+    return result
+
+
+preamble = """// {0} for vips operations
+// {1}
+// this file is generated automatically, do not edit!
+"""
 
-    print 'VImage::%s(' % nickname,
+now = datetime.datetime.now().strftime('%a %d %b %H:%M:%S %Y')
 
-    gen_arg_list(op, required)
 
-    print ')',
-    if this != None:
-        print 'const',
-    print
+def generate_operators(filename, declarations_only=False):
+    all_nicknames = []
 
-    print '{'
-    if result != None:
-        print '    %s %s;' % (get_ctype(result), cppize(result.name))
-        print ''
+    def add_nickname(gtype, a, b):
+        nickname = nickname_find(gtype)
+        try:
+            # can fail for abstract types
+            op = Operation.new_from_name(nickname)
 
-    print '    call( "%s"' % nickname,
+            # we are only interested in non-deprecated operations
+            if (op.get_flags() & _OPERATION_DEPRECATED) == 0:
+                all_nicknames.append(nickname)
+        except Error:
+            pass
 
-    first = True
-    for prop in all_required:
-        if first:
-            print ','
-            print '        (options ? options : VImage::option())',
-            first = False
+        type_map(gtype, add_nickname)
 
-        print '->'
-        print '           ',
-        if prop == this:
-            print 'set( "%s", *this )' % prop.name,
-        else:
-            flags = op.get_argument_flags(prop.name)
-            arg = cppize(prop.name)
-            if flags & Vips.ArgumentFlags.OUTPUT and prop == result:
-                arg = '&' + arg
+        return ffi.NULL
 
-            print 'set( "%s", %s )' % (prop.name, arg),
+    type_map(type_from_name('VipsOperation'), add_nickname)
 
-    print ');'
+    # TODO: Should we add the 'missing' extract_area synonym by hand?
+    # all_nicknames.append('crop')
 
-    if result != None:
-        print ''
-        print '    return( %s );' % cppize(result.name)
+    # make list unique and sort
+    all_nicknames = list(set(all_nicknames))
+    all_nicknames.sort()
 
-    print '}'
-    print ''
+    print('Generating {0} ...'.format(filename))
```

Comment:
This script is usually run by `Makefile.am` if `vips-operators.cpp` is missing. It's probably simpler to do the date stuff there. Or remove the stuff from the Makefile if we're going to get maintainers to run the script by hand.